### PR TITLE
.NET: BackgroundImage tiling

### PR DESCRIPTION
**Pull request by carlos-zamora**
Implementation for #1362 (Spec #477 ).
Please double check XmlSerialization.

# Parsing, (De)Serialization
- `AdaptiveBackgroundImage.cs`: pretty standard addition of new object
- `AdaptiveBackgroundImageConverter.cs`: special converter to handle backcompat + ducktyping use of BackgroundImage
-- `backgroundImage: <string>` vs `backgroundImage: { <properties> }`

# WPF
- Updated `SetBackgroundImageSource(...)` to...
-- take in `AdaptiveBackgroundImage` instead of `Uri`
-- handle control flow for different `BackgroundImageModes`
- Added `SetBackgroundImageSource(...)` calls to `ContainerRenderer` and `ColumnRenderer`

# HTML
- Created `ApplyBackgroundImage(...)` to handle control flow for `BackgroundImageModes`
- Added `ApplyBackgroundImage(...)` calls to `Container` and `Column` rendering

### CODE REVIEW

File: source/dotnet/Library/AdaptiveCards/AdaptiveBackgroundImageConverter.cs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        public List<AdaptiveWarning> Warnings { get; set; } = new List<AdaptiveWarning>();
+
+        public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
+        {
+            AdaptiveBackgroundImage bi = (AdaptiveBackgroundImage) value;
+            if (bi.HasDefaultValues())
+            {
+                writer.WriteValue(bi.BackgroundImageString);
+            }
+            else
```
Comment (khouzam) [inline source/dotnet/Library/AdaptiveCards/AdaptiveBackgroundImageConverter.cs]: Should this check if BackgroundImageString is null or empty and then not serialize anything?

File: source/dotnet/Test/AdaptiveCards.Test/SerializationTests.cs [hunk 2]
```diff
@@ -20,7 +20,7 @@ public void TestCardsSerializeInTheCorrectOrder()
             card.Version = "1.0";
             card.FallbackText = "Fallback Text";
             card.Speak = "Speak";
-            card.BackgroundImage = new Uri("http://adaptivecards.io/content/cats/1.png");
+            card.BackgroundImage = new AdaptiveBackgroundImage("http://adaptivecards.io/content/cats/1.png");
             card.Body.Add(new AdaptiveTextBlock { Text = "Hello" });
```
Comment (khouzam) [inline source/dotnet/Test/AdaptiveCards.Test/SerializationTests.cs]: Can we get some test coverage for the other properties with the background image serialization as well as for the containers?

File: source/dotnet/Library/AdaptiveCards.Rendering.Wpf/Helpers/ImageExtensions.cs [hunk 3]
```diff
@@ -142,12 +142,47 @@ public static async void SetBackgroundSource(this Grid grid, Uri uri, AdaptiveRe
...
+                            ViewportUnits = BrushMappingMode.Absolute
+                        };
+                        break;
+                    case AdaptiveBackgroundImageMode.RepeatHorizontally:
+                        grid.Background = new ImageBrush(bi)
+                        {
+                            TileMode = TileMode.FlipY,
+                            Stretch = Stretch.Uniform,
+                            AlignmentY = (AlignmentY) adaptiveBackgroundImage.VerticalAlignment,
+                            Viewport = new Rect(0, 0, bi.PixelWidth, grid.ActualHeight+1),
```
Comment (almedina-ms) [inline source/dotnet/Library/AdaptiveCards.Rendering.Wpf/Helpers/ImageExtensions.cs]: >grid.ActualHeight+1 [](start = 69, length = 19)

should be grid.ActualHeight +1

File: source/dotnet/Library/AdaptiveCards.Rendering.Wpf/Helpers/ImageExtensions.cs [hunk 4]
```diff
@@ -142,12 +142,47 @@ public static async void SetBackgroundSource(this Grid grid, Uri uri, AdaptiveRe
...
+                            ViewportUnits = BrushMappingMode.Absolute
+                        };
+                        break;
+                    case AdaptiveBackgroundImageMode.RepeatVertically:
+                        grid.Background = new ImageBrush(bi)
+                        {
+                            TileMode = TileMode.FlipX,
+                            Stretch = Stretch.Uniform,
+                            AlignmentX = (AlignmentX)adaptiveBackgroundImage.HorizontalAlignment,
+                            Viewport = new Rect(0, 0, grid.ActualWidth+1, bi.PixelWidth),
```
Comment (almedina-ms) [inline source/dotnet/Library/AdaptiveCards.Rendering.Wpf/Helpers/ImageExtensions.cs]: >grid.ActualWidth+1 [](start = 54, length = 18)

grid.ActualWidth + 1

File: source/dotnet/Library/AdaptiveCards/AdaptiveBackgroundImageConverter.cs [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        {
+            JToken backgroundImageJSON = JToken.Load(reader);
+
+            // Handle BackgroundImage as a string (BackCompat)
+            if (backgroundImageJSON.Type == JTokenType.String)
+            {
+                return new AdaptiveBackgroundImage(backgroundImageJSON.Value<string>());
+            }
+
+            // backgroundImage is an object (Modern)
```
Comment (almedina-ms) [inline source/dotnet/Library/AdaptiveCards/AdaptiveBackgroundImageConverter.cs]: nit: extra line

File: source/dotnet/Library/AdaptiveCards/AdaptiveBackgroundImageConverter.cs [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+            else if (backgroundImageJSON.Type == JTokenType.Object)
+            {
+                return backgroundImageJSON.ToObject<AdaptiveBackgroundImage>();
+            }
+
+            else
```
Comment (almedina-ms) [inline source/dotnet/Library/AdaptiveCards/AdaptiveBackgroundImageConverter.cs]: nit: extra line

File: source/dotnet/Test/AdaptiveCards.Test/SerializationTests.cs (hunk 2 above)
Comment (carlos-zamora) [reply source/dotnet/Test/AdaptiveCards.Test/SerializationTests.cs]: Testing them with the JSON tests in #2316.

File: source/dotnet/Library/AdaptiveCards/AdaptiveBackgroundImageConverter.cs [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        public override bool CanWrite => true;
+
+        public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
+        {
+            JToken backgroundImageJSON = JToken.Load(reader);
+
+            // Handle BackgroundImage as a string (BackCompat)
+            if (backgroundImageJSON.Type == JTokenType.String)
+            {
+                return new AdaptiveBackgroundImage(backgroundImageJSON.Value<string>());
```
[the remark this answers is not in the document]
Comment (carlos-zamora) [reply source/dotnet/Library/AdaptiveCards/AdaptiveBackgroundImageConverter.cs]: This somehow doesn't happen. In master, when the card has `"backgroundImage": ""`, the backgroundImage object is just set to null. Here, it still creates a BackgroundImage object with default fields but the string is also set to "". Surprisingly, no exception is thrown and everything just works fine.

File: source/dotnet/Test/AdaptiveCards.Test/SerializationTests.cs (hunk 2 above)
Comment (carlos-zamora) [reply source/dotnet/Test/AdaptiveCards.Test/SerializationTests.cs]: Added test on line 425.

---
In reply to: [250430169](https://github.com/Microsoft/AdaptiveCards/pull/2234#discussion_r250430169) [](ancestors = 250430169)

Comment (matthidinger) [reply source/dotnet/Test/AdaptiveCards.Test/SerializationTests.cs]: This is a breaking change to the API, we should be mindful of these. I pushed a fix that adds implicit conversion from Uri to the new AdaptiveBackgroundImage image type

**Comment by matthidinger** [comment]
I fixed the test for XmlSerialization, it should run consistently now as part of our normal unit test verification. Had to make a minor change to the PixelHeight property. Pushed everything now

File: source/dotnet/Library/AdaptiveCards/AdaptiveBackgroundImageConverter.cs [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+namespace AdaptiveCards
+{
+    public class AdaptiveBackgroundImageConverter : JsonConverter, ILogWarnings
+    {
+        public List<AdaptiveWarning> Warnings { get; set; } = new List<AdaptiveWarning>();
+
+        public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
+        {
+            AdaptiveBackgroundImage bi = (AdaptiveBackgroundImage) value;
+            if (bi.UrlString != null && bi.UrlString != "")
```
Comment (khouzam) [inline source/dotnet/Library/AdaptiveCards/AdaptiveBackgroundImageConverter.cs]: >bi.UrlString != null && bi.UrlString != "" [](start = 16, length = 42)

`string.IsNullOrEmpty(bi.UrlString)` is slightly easier

**Pull request merged**